compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala

```diff
@@ -48,12 +61,103 @@ class ExtractSemanticDB extends Phase:
 
   override def run(using Context): Unit =
     val unit = ctx.compilationUnit
-    val extractor = Extractor()
-    extractor.extract(unit.tpdTree)
-    ExtractSemanticDB.write(unit.source, extractor.occurrences.toList, extractor.symbolInfos.toList, extractor.synthetics.toList)
+    if (phaseMode == ExtractSemanticDB.PhaseMode.PostTyper)
+      val extractor = ExtractSemanticDB.Extractor()
+      extractor.extract(unit.tpdTree)
+      ExtractSemanticDB.write(unit.source, extractor.occurrences.toList, extractor.symbolInfos.toList, extractor.synthetics.toList)
+    else
+      val warnings = ctx.reporter.allWarnings.collect {
+        case w if w.pos.source == ctx.source => w.toSemanticDiagnostic
+      }
+      if (warnings.nonEmpty)
+        ExtractSemanticDB.appendDiagnostics(unit.source, warnings)
+end ExtractSemanticDB
+
+object ExtractSemanticDB:
+  import java.nio.file.Path
+  import java.nio.file.Files
+  import java.nio.file.Paths
+
+  val phaseNamePrefix: String = "extractSemanticDB"
+  val description: String = "extract info into .semanticdb files"
+
+  enum PhaseMode:
+    case PostTyper
+    case PostInlining
+
+  class PostTyper extends ExtractSemanticDB(PhaseMode.PostTyper, "PostTyper")
+
+  class PostInlining extends ExtractSemanticDB(PhaseMode.PostInlining, "PostInlining")
+
+  private def semanticdbTarget(using Context): Option[Path] =
+    Option(ctx.settings.semanticdbTarget.value)
+      .filterNot(_.isEmpty)
+      .map(Paths.get(_))
+
+  private def outputDirectory(using Context): AbstractFile = ctx.settings.outputDir.value
+
+  private def absolutePath(path: Path): Path = path.toAbsolutePath.normalize
+
+  private def write(
+    source: SourceFile,
+    occurrences: List[SymbolOccurrence],
+    symbolInfos: List[SymbolInformation],
+    synthetics: List[Synthetic],
+  )(using Context): Unit =
+    val outpath = semanticdbPath(source)
+    Files.createDirectories(outpath.getParent())
+    val doc: TextDocument = TextDocument(
+      schema = Schema.SEMANTICDB4,
+      language = Language.SCALA,
+      uri = Tools.mkURIstring(Paths.get(relPath(source))),
+      text = "",
+      md5 = internal.MD5.compute(String(source.content)),
+      symbols = symbolInfos,
+      occurrences = occurrences,
+      synthetics = synthetics,
+    )
+    val docs = TextDocuments(List(doc))
+    val out = Files.newOutputStream(outpath)
+    try
+      val stream = internal.SemanticdbOutputStream.newInstance(out)
+      docs.writeTo(stream)
+      stream.flush()
+    finally
+      out.close()
+  end write
+
+  private def appendDiagnostics(
+    source: SourceFile,
+    diagnostics: Seq[Diagnostic]
+  )(using Context): Unit =
+    val path = semanticdbPath(source)
+    Using.Manager { use =>
+      val in = use(Files.newInputStream(path))
+      val sin = internal.SemanticdbInputStream.newInstance(in)
+      val docs = TextDocuments.parseFrom(sin)
```

Comment:
this takes about 62% of the time of this method, I wonder if there's a cheaper way to directly append the diagnostics

Reply:
No, I don't think there's a cheaper alternative way, since we don't know the protobuf structure / where in the binary to write without parsing it 😞

Reply:
From a protobuf perspective, it is possible to concatenate the serialized representation of a [`TextDocument`](https://github.com/scalameta/scalameta/blob/2e0ad9e6a9ed8ba16b15094c61e3b06936af11ac/semanticdb/semanticdb/semanticdb.proto#L21-L32) containing only the diagnostics to the serialized representation of the existing one, since encoding does not rely on ordering.

Unfortunately, semanticdb files are [`TextDocumentS`](https://github.com/scalameta/scalameta/blob/2e0ad9e6a9ed8ba16b15094c61e3b06936af11ac/semanticdb/semanticdb/semanticdb.proto#L17-L19) messages (effectively containing one `TextDocument`), and considering [submessages are length-encoded](https://protobuf.dev/programming-guides/encoding/#embedded), concatenating a `TextDocument` to a `TextDocuments` would work only if the length of the original `TextDocument` was updated. Code generated by ScalaPB does not allow that without deserializing the submessage :disappointed: 

I can think of a workaround though: **set the contatenation of the serialized representations of the diagnostic-free and the diagnostic-only `TextDocument` [through an unknown field](https://github.com/scalapb/ScalaPB/blob/b3d0f22c1313eb7f1d2e12d3047b59b83eee3c17/scalapb-runtime/src/main/scala/scalapb/UnknownFieldSet.scala#L55) on `TextDocuments`**. It should work as unknown fields can apparently shadow known fields ([there is no check on that](https://github.com/scalapb/ScalaPB/blob/b3d0f22c1313eb7f1d2e12d3047b59b83eee3c17/scalapb-runtime/src/main/scala/scalapb/UnknownFieldSet.scala#L22-L26)), but I haven't tried it myself :smile: 


Note that
1. The serialized representation of the diagnostic-free `TextDocument` would need to be retrieved using an unknown field set to avoid parsing, for example by loading the semanticdb file [through an `Empty` message](https://github.com/scalapb/ScalaPB/blob/b3d0f22c1313eb7f1d2e12d3047b59b83eee3c17/scalapb-runtime/src/main/scalajvm/com/google/protobuf/empty/Empty.scala#L20).
1. The unknown field feature [would need to be enabled first](https://github.com/tanishiking/semanticdb-for-scala3/blob/08b0e587dbbbea98a758b31ae028ffb20c50551f/generator/src/main/protobuf/scalameta.proto#L8).

It sounds pretty complicated but it should be only a couple of extra lines.

Reply:
Oh, thanks @bjaglin!
I'll delve into the things you explained :)

Reply:
so I'll resolve this comment as we are not going to do it in this PR, but worth revisiting